google/cloud/storage/blob.py

```diff
@@ -1314,6 +1314,11 @@ def _do_resumable_upload(
         :returns: The "200 OK" response object returned after the final chunk
                   is uploaded.
         """
+        if size is not None:
```

Comment:
Yes, there is a bug, the culprit  of error is [end_byte >= total_bytes](https://github.com/googleapis/google-resumable-media-python/blob/master/google/resumable_media/_upload.py#L864)  line where we set `total_byte = size`, but `end_byte` is the total length of a stream so end_byte always greater than total_byte and raise the error. Here we need to alter the stream and create a stream with the same size as total_bytes.


Have two ways to handle this error:

- Handled in current PR as `size` parameter manged in [_do_multipart](https://github.com/googleapis/python-storage/blob/304024db1a52d03b18b7837c994893e7fd56a822/google/cloud/storage/blob.py#L1050-L1056) method.

- We can handle in [_prepare_initiate_request](https://github.com/googleapis/google-resumable-media-python/blob/b306c780311133bae38f1dbd3f030244f75e32c2/google/resumable_media/_upload.py#L373) method where we set `size` to `total_bytes`

Reply:
I'm digging in a little bit more because a bit confused still. Could you do me a favor and log requests going to GCS because I think this should be handled in: https://github.com/googleapis/google-resumable-media-python/blob/b306c780311133bae38f1dbd3f030244f75e32c2/google/resumable_media/_upload.py#L599-L610

total_bytes are passed to the resumable media but it doesn't seem that the upload completes knowing the total_bytes when transmitting chunks.

The issue with the current PR is that it will hold large amount of data in-memory and we don't want to do that for resumable media. For multi-part it's okay because there's a fixed size that's understood. 

Could you dig in there some more?

Reply:
Here is the trace back as far as i dig in:

First where hit the [initiate](https://github.com/googleapis/google-resumable-media-python/blob/master/google/resumable_media/requests/upload.py#L295) method and call the [_prepare_initiate_request](https://github.com/googleapis/google-resumable-media-python/blob/master/google/resumable_media/requests/upload.py#L338) it set the header `x-upload-content-length': total_bytes` , where total_byte is `size` or length of `stream` and `data` is `metadata` which has only file name, In response it gives the 200 with blank text.

Then called [upload.transmit_next_chunk](https://github.com/googleapis/google-resumable-media-python/blob/master/google/resumable_media/requests/upload.py#L356) method  so through `_prepare_request` call it reach to the method [get_next_chunk](https://github.com/googleapis/google-resumable-media-python/blob/b306c780311133bae38f1dbd3f030244f75e32c2/google/resumable_media/_upload.py#L812) where we have check of [end_byte>total_byte](https://github.com/googleapis/google-resumable-media-python/blob/b306c780311133bae38f1dbd3f030244f75e32c2/google/resumable_media/_upload.py#L864) where `end_byte` is length of stream and `total_length` is size set by user then our library throws the value error from [here](https://github.com/googleapis/google-resumable-media-python/blob/b306c780311133bae38f1dbd3f030244f75e32c2/google/resumable_media/_upload.py#L865-L866)

Reply:
In each transmit_next_chunk(), there should be context in the request on the offset and total byte size of the object. https://cloud.google.com/storage/docs/performing-resumable-uploads#multiple-chunks

If total_bytes is correct then Cloud Storage will complete the upload. I recommend reviewing request logs per: https://medium.com/@franknatividad/debug-cloud-storage-client-libraries-f50f46ce7f6

It might be that range isn't being set properly that's why the code is hitting the value error that you pointed to.

Reply:
Yes, that's what i was elaborating,  get_next_chunk method has a line which calculate [end_byte](https://github.com/googleapis/google-resumable-media-python/blob/b306c780311133bae38f1dbd3f030244f75e32c2/google/resumable_media/_upload.py#L843) directly from stream not through the total_bytes. This happens before the first request hit with the chunk to the server. 

So i have two option on top of my mind, need your favor:

-  Set the total_byte value to the end_byte if found any difference, but then there is not point of value error which checks the `end_byte>total_byte`.

- Need to edit the stream before it sets the end_byte.

Reply:
@frankyn  any suggestion here, i didn't find any other solution apart from which i have mentioned in comment above.

Reply:
@frankyn Could you suggest something here? I haven't found any other possible solution apart from which i have mentioned in comment above.

Reply:
Apologies @HemangChothani, I've been working my way back to this. Looking at it now.

Reply:
Thanks for your patience @HemangChothani, I was buliding context and understand your points.

I like your solution 1 in this case as it's cleaner but does need to drop the value error (which isn't bad), but not sure why this was implemented this way in the first place. 

IIUC, you'll do the following: https://github.com/googleapis/google-resumable-media-python/blob/dc04017f5e5c13f825ae96de2ba1f59762cc1982/google/resumable_media/_upload.py#L841-L869

I don't see a reason why not to right now. Can you continue with this path and close this PR?

Reply:
@frankyn Thanks for your guidance , Closing this PR and i will open a new PR in google-resumable-media soon.